samples/nrf9160/serial_lte_modem/src/main.c

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#define LOG_DOMAIN app
+#define LOG_LEVEL CONFIG_THIN_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_DOMAIN);
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <uart.h>
+#include <string.h>
+#include <net/socket.h>
+#include <init.h>
+#include <bsd.h>
+#include <lte_lc.h>
+#include <net/mqtt.h>
+
+#include <hal/nrf_gpio.h>
+#include <hal/nrf_power.h>
+#include <hal/nrf_regulators.h>
+
+#include "th_test.h"
+#include "th_connect.h"
+#if defined(CONFIG_LTE_LINK_CONTROL)
+#include "th_modem_service.h"
+#endif
+#if defined(CONFIG_THIN_MQTT_SERVICE)
+#include "th_mqtt_service.h"
+#endif
+#if defined(CONFIG_THIN_LWM2M_SERVICE)
+#include "th_lwm2m_service.h"
+#endif
+#if defined(CONFIG_THIN_GPS_SERVICE)
+#include "th_gps_service.h"
+#endif
+#if defined(CONFIG_THIN_SNTP_SERVICE)
+#include "th_sntp_service.h"
+#endif
+
+#define INVALID_DESCRIPTOR	-1
+
+#define THREAD_STACK_SIZE	(CONFIG_THIN_MODEM_SOCKET_BUF_SIZE + 512)
+#define THREAD_PRIORITY		K_PRIO_PREEMPT(CONFIG_THIN_MODEM_THREAD_PRIO)
+
+static int at_socket_fd = INVALID_DESCRIPTOR;
+static struct pollfd fds[1];
+static int nfds;
+static u8_t at_buf[CONFIG_THIN_MODEM_UART_BUF_SIZE];
+static size_t at_buf_len;
+static struct k_work at_cmd_send_work;
+static struct k_thread socket_thread;
+static K_THREAD_STACK_DEFINE(socket_thread_stack, THREAD_STACK_SIZE);
+static struct k_mutex socket_mutex;
+static bool at_cmd_responded;
+
+static void at_cmd_send(struct k_work *work)
+{
+	int bytes_sent;
+
+	ARG_UNUSED(work);
+
+	k_mutex_lock(&socket_mutex, K_FOREVER);
+	bytes_sent = send(at_socket_fd, at_buf, at_buf_len, 0);
+	k_mutex_unlock(&socket_mutex);
+
+	if (bytes_sent <= 0) {
+		LOG_WRN("Could not send AT command to modem: %d", bytes_sent);
+	}
+}
+
+static void at_socket_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	u8_t at_read_buff[CONFIG_THIN_MODEM_SOCKET_BUF_SIZE] = {0};
+	int err;
+	int r_bytes;
+
+	ARG_UNUSED(arg1);
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	while (1) {
+		/* Poll the socket for incoming data. */
+		err = poll(fds, nfds, K_FOREVER);
+		if (err < 0) {
+			LOG_ERR("Poll error: %d", err);
+			continue;
+		}
+
+		k_mutex_lock(&socket_mutex, K_FOREVER);
```

Comment:
Did you encounter any issues with parallel send/recv so that we need a socket here? I'm asking, because theoretically TX/RX path should work in parallel, even when called from different threads, and a need for a mutex complicates the design a bit (for instance it prevents us from using a blocking socket here, which would be much simpler as we have a separate thread only for socket).

Reply:
No I did not. I have not thought it over. Originally I want to prevent concurrent TX/RX over the UART interface, not the socket interface. I'll remove it for now.